### PR TITLE
feat: create protected routes 

### LINKED ISSUE #38: [frontend] create protected route
create a component to check the user role and return if he can pass to rout or not

### PR DESCRIPTION
**Pull request by mo7amedehab97**
feat: create protected routes

### CODE REVIEW

**Review by braaAwni — DISMISSED** [review]

**Review by nabilramy — DISMISSED** [review]

File: client/src/ProtectedRoutes/ProtectedRoutes.jsx [hunk 1]
```diff
@@ -0,0 +1,37 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import {
+  Navigate,
+} from 'react-router-dom';
+import { useAuth } from '../Context/AuthContext';
+
+export function ProtectedAdmin({ children }) {
+  const { user } = useAuth();
+  if (!user.user) {
```
Comment (aaamra) [inline client/src/ProtectedRoutes/ProtectedRoutes.jsx]: you should check the role for the admin

File: client/src/ProtectedRoutes/ProtectedRoutes.jsx [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    return <Navigate to="/dashboard" replace />;
+  }
+  return children;
+}
+export function ProtectedUser({ children }) {
+  const { user } = useAuth();
+  if (user.user.role === 0) {
```
Comment (aaamra) [inline client/src/ProtectedRoutes/ProtectedRoutes.jsx]: and here you just want to check if there is a user in the context
not the user role

File: client/src/ProtectedRoutes/ProtectedRoutes.jsx [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    return <Navigate to="/" replace />;
+  }
+  return children;
+}
+export function ProtectedEngineer({ children }) {
+  const { user } = useAuth();
+  if (user.user.role === 2) {
```
Comment (aaamra) [inline client/src/ProtectedRoutes/ProtectedRoutes.jsx]: as I told you it's not readable to keep the roles as numbers. you can define constants in the useAuth hook to make it more readable

**Review by aaamra — CHANGES_REQUESTED** [review]

File: client/src/ProtectedRoutes/ProtectedRoutes.jsx (hunk 1 above)
Comment (aaamra) [reply client/src/ProtectedRoutes/ProtectedRoutes.jsx:10]: in the useAuth hook change it and store just the user in the user state
you are calling `user.user`
keep it simple and make it like this 
```suggestion
  if (!user) {
```

**Review by aaamra — CHANGES_REQUESTED** [review]

File: client/src/ProtectedRoutes/ProtectedRoutes.jsx (hunk 3 above)
Comment (aaamra) [reply client/src/ProtectedRoutes/ProtectedRoutes.jsx]:
```suggestion
  if (user.role === ENG_ROLE) {
```

**Review by aaamra — APPROVED** [review]

**Review by nabilramy — APPROVED** [review]

**Review by KAHMOOSHA — APPROVED** [review]

**Review by braaAwni — APPROVED** [review]

**Pull request merged**